base/applications/rapps/asyncinet.cpp

```diff
@@ -0,0 +1,258 @@
+            {
+                // this is most likely not an error. when canceling, using closed/invalid handle may lead to this
+                break;
+            }
+            // fall-down to handle error
+        default:
+            // something went wrong
+            AsyncInetPerformCallback(AsyncInet, ASYNCINET_ERROR, 0, (LPARAM)(AsyncResult->dwError));
+            break;
+        }
+        break;
+    }
+
+    }
+    return;
+}
+
+
+pASYNCINET AsyncInetDownload(LPCWSTR lpszAgent,
+    DWORD   dwAccessType,
+    LPCWSTR lpszProxy,
+    LPCWSTR lpszProxyBypass,
+    LPCWSTR lpszUrl,
+    BOOL bAllowCache,
+    ASYNCINET_CALLBACK Callback,
+    VOID* Extension
+    )
+{
+    pASYNCINET AsyncInet = NULL;
+    BOOL bSuccess = FALSE;
+    DWORD InetOpenUrlFlag = 0;
+    INTERNET_STATUS_CALLBACK OldCallbackFunc;
+
+    AsyncInet = (pASYNCINET)HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, sizeof(ASYNCINET));
+    if (!AsyncInet)goto cleanup;
+
+    AsyncInet->Callback = Callback;
+    AsyncInet->Extension = Extension;
+
+    AsyncInet->hInternet = InternetOpenW(lpszAgent, dwAccessType, lpszProxy, lpszProxyBypass, INTERNET_FLAG_ASYNC);
+    if (!AsyncInet->hInternet)goto cleanup;
+
+    OldCallbackFunc = InternetSetStatusCallbackW(AsyncInet->hInternet, AsyncInetStatusCallback);
+    if (OldCallbackFunc == INTERNET_INVALID_STATUS_CALLBACK)goto cleanup;
+
+    InetOpenUrlFlag |= INTERNET_FLAG_KEEP_CONNECTION | INTERNET_FLAG_PASSIVE;
+    if (!bAllowCache)
+    {
+        InetOpenUrlFlag |= INTERNET_FLAG_DONT_CACHE | INTERNET_FLAG_PRAGMA_NOCACHE;
+    }
+    AsyncInet->hInetFile = InternetOpenUrlW(AsyncInet->hInternet, lpszUrl, 0, 0, InetOpenUrlFlag, (DWORD_PTR)AsyncInet);
+    if (!AsyncInet->hInetFile)
+    {
+        if (GetLastError() != ERROR_IO_PENDING)
+        {
+            goto cleanup;
+        }
+    }
+    else
+    {
+        // TODO: If I remember it correctly, sometimes the file is cache before
+        // thus may lead to InternetOpenUrlW return immediately with the handle.
+        // more investigation required. And if it's true, it should be handled correctly.
+    }
+
+    bSuccess = TRUE;
+
+
+cleanup:
+
+    if (!bSuccess)
+    {
+        if (AsyncInet)
+        {
+            if (AsyncInet->hInternet)
+            {
+                InternetCloseHandle(AsyncInet->hInternet);
+                AsyncInet->hInternet = NULL;
+            }
+            if (AsyncInet->hInetFile)
+            {
+                InternetCloseHandle(AsyncInet->hInetFile);
+                AsyncInet->hInetFile = NULL;
+            }
+            HeapFree(GetProcessHeap(), 0, AsyncInet);
+            AsyncInet = NULL;
+        }
+    }
+
+    return AsyncInet;
+}
+
+
+BOOL AsyncInetCancel(pASYNCINET AsyncInet)
+{
+    if (!AsyncInet)
+    {
+        return FALSE;
+    }
+
+    AsyncInet->bIsCancelled = TRUE;
+
+    // AsyncInet may got freed while handling INTERNET_STATUS_HANDLE_CLOSING.
+    // so store the handle first, then close them
```

Comment:
That seems like a race condition then?

Reply:
I think yes. we don't know which will happen first, download complete or download cancelled

Reply:
Both `AsyncInetDownload` and this function close these handles, but one of them cleans up the object.
This seems weird?